### PR TITLE
Documentation improvements

### PR DESCRIPTION
**Pull request by kepta**
This improves the documentation of the configuration object.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -314,52 +315,95 @@ module.exports = {
 };
 ```
 
-## Configuration
+## Underreact configuration file
 
-**No configuration is necessary to get started.** On most production projects you'll want to set at least a few of the "Basic configuration" options. "Advanced configuration" options shouldn't be necessary in *most* cases ... but you never know.
+Underreact expects a `underreact.config.js` file to exist at the root of your project. Please note that **No configuration is necessary to get started.** On most production projects you'll want to set at least a few of the [configuration object](#configuration-object-properties) properties.
```
Comment (davidtheclark) [inline README.md]: A couple typos in here:

expects a > expect an
No configuration > no configuration

File: README.md [hunk 2]
```diff
@@ -314,52 +315,95 @@ module.exports = {
...
 
-### siteBasePath
+Your `underreact.config.js` can look like either of the below:
 
-Type: `string`. Default: `''`.
+- **Exporting a Function (Preferred)**: You can export a function which would then be used as a factory method for your [configuration object](#configuration-object-properties). This function is called the following parameter properties of an object:
```
Comment (davidtheclark) [inline README.md]: We could lowercase Function and Preferred (and Object below).

Comment (davidtheclark) [inline README.md]: called the > called with the

File: README.md [hunk 3]
```diff
@@ -314,52 +315,95 @@ module.exports = {
...
 
-Root-relative path to the base directory on the domain where the site will be deployed.
+```javascript
+// underreact.config.js
+/**
+ * @param {Object} opts 
+ * @param {Webpack} opts.webpack - The Webpack object.
```
Comment (davidtheclark) [inline README.md]: Whether here or elsewhere, we should explain what this is and why you'd use it.

File: README.md [hunk 4]
```diff
@@ -314,52 +315,95 @@ module.exports = {
...
+ * @param {'start'|'build'|'serve-static'} opts.command - The current command Underreact is following.
+ * @param {'production'|'development'} opts.mode - The current mode of Underreact.
+ */
```
Comment (davidtheclark) [inline README.md]: Adding a `@returns` annotation would complete the JSDoc comment.

File: README.md [hunk 5]
```diff
@@ -314,52 +315,95 @@ module.exports = {
...
+module.exports = function({ webpack, command, mode }) {
+  return {/*Underreact configuration object*/}
+}
+```
 
-There's a good chance your app isn't at the root of your domain. So this option represents the path of your site *within* that domain.
+- **Exporting an Object**: You can also directly export the [configuration object](#configuration-object-properties). This is a great way to start experimenting with Underreact's configuration. For example in the code below we simply want to modify the `siteBasePath`:
```
Comment (davidtheclark) [inline README.md]: We have a reason why we might want to export an object, but I don't see any reasons why we might want to export a function (even though it's preferred). Some explanation of that would help.

File: README.md [hunk 6]
```diff
@@ -314,52 +315,95 @@ module.exports = {
...
-For example, if your app is at `https://www.special.com/ketchup/*`, you should set `siteBasePath: ketchup`.
+```javascript
+// underreact.config.js
+module.exports = {
+   siteBasePath: 'fancy'
+}
+```
 
-### stylesheets
+## Configuration Object Properties
```
Comment (davidtheclark) [inline README.md]: The convention I've been going with is to capitalize only the first word of headings (or, of course, other words that need to be capitalized, like names).

File: README.md [hunk 7]
```diff
@@ -314,52 +315,95 @@ module.exports = {
...
+Set to `true` if you want to use HTML5 History for client-side routing (as opposed to hash routing). This configures the development server to fall back to `index.html` when you request nested routes.
 
-Assets referenced by your stylesheets will be hashed and copied to the [outputDirectory](#outputdirectory), unless they are absolute URLs.
+This is `false` by default because it should only be *intentionally* turned on, when you know you're going to configure your server to allow for HTML5 History—powered client-side routing.
 
-### vendorModules
+### htmlSource
 
-Type: `Array<string>`. Default: `[]`.
+Type: `string`. Absolute path, please. Default: `${project-directory}/src/html.js`.
```
Comment (davidtheclark) [inline README.md]: Do you not want to change this to a `require` after all?

File: README.md [hunk 8]
```diff
@@ -314,52 +315,95 @@ module.exports = {
...
 
-Identifiers of npm modules that you want to be added to the vendor bundle.
-The purpose of the vendor bundle is to deliberately group dependencies that change relatively infrequently — so this bundle will stay cached for longer than the others.
+The path to your HTML source file. For more information, read [Defining your HTML](#defining-your-html).
```
Comment (davidtheclark) [inline README.md]: A convention I've been trying to follow is to wrap section titles in quotation marks.

File: README.md [hunk 9]
```diff
@@ -314,52 +315,95 @@ module.exports = {
...
-### htmlSource
+Type `string`. Absolute path, please. Default: `${project-directory}/_underreact-site/`.
 
-Type: `string`. Absolute path, please. Default: `${project-directory}/src/html.js`.
+The directory where files should be written.
 
-The path to your HTML source file. For more information, read [Defining your HTML](#defining-your-html).
+You'll want to ignore this directory with `.gitignore`, `.eslintignore`, etc.
+
+In the default value, `project-directory` refers to the directory of your `underreact.config.js` file, or the current working directory.
```
Comment (davidtheclark) [inline README.md]: "or the current working directory" when?

File: README.md [hunk 10]
```diff
@@ -314,52 +315,95 @@ module.exports = {
...
+
+### publicDirectory
+
+Type `string`. Absolute path, please. Default: `${project-directory}/src/public/`.
+
+Any files you put into this directory will be copied, without processing, into the [outputDirectory](#outputdirectory).
+You can put images, favicons, data files, anything else you want in here.
```
Comment (davidtheclark) [inline README.md]: data files, anything > data files, and anything

File: README.md [hunk 11]
```diff
@@ -368,6 +412,26 @@ Type: `Array<Function>`. Default: [Autoprefixer](https://github.com/postcss/auto
...
 
+### siteBasePath
+
+Type: `string`. Default: `''`.
+
+Root-relative path to the base directory on the domain where the site will be deployed.
+
+There's a good chance your app isn't at the root of your domain. So this option represents the path of your site *within* that domain.
+
+For example, if your app is at `https://www.special.com/ketchup/*`, you should set `siteBasePath: ketchup`.
```
Comment (davidtheclark) [inline README.md]: ketchup > 'ketchup' (quotes)

File: README.md [hunk 12]
```diff
@@ -368,6 +412,26 @@ Type: `Array<Function>`. Default: [Autoprefixer](https://github.com/postcss/auto
...
+
+### stylesheets
+
+Type: `Array<string>`. Absolute paths, please. Default: `[]`.
+
+An array of filenames pointing to stylesheets that you want to include in your site.
+
+These will be processed by PostCSS with [Autoprefixer](https://github.com/postcss/autoprefixer) and any other [postcssPlugins](#postcssplugins) that you specify; concatenated in the order specified; and added to the `<head>` of your document.
```
Comment (davidtheclark) [inline README.md]: Currently they are not automatically added to the `<head>`, right?

File: README.md [hunk 13]
```diff
@@ -403,46 +467,11 @@ Type: `config => transformedConfig`. Default `x => x` (identify function).
...
-
-The directory where Underreact assets will be placed, relative to the site's root.
-
-By default, for example, the main JS chunk will be written to `underreact-assets/main.chunk.js`.
-
-It's important to know about this value so you can set up caching and other asset configuration on your server.
+### vendorModules
 
-### port
+Type: `Array<string>`. Default: `[]`.
```
Comment (davidtheclark) [inline README.md]: Should the default be `['react', 'react-dom']`?

Comment (kepta) [reply README.md]: Yes it is, however I am a bit worried about the implementation as we concat this to an internal array of vendor. I wonder if that is of any real concern to user ?

File: README.md (hunk 7 above)
Comment (kepta) [reply README.md]: Yes I do, but I got stalled after worrying about how do we watch changes to it.

Right now, it is easy as we simply have a path we can watch to, but I wonder if there is a good way to watch to do that for a real javascript function.

File: README.md (hunk 5 above)
Comment (kepta) [reply README.md]: I tried to convince myself about why exporting a function is a preferred way, but failed. I feel both of the methods are equal and the object method tad more simple.

I have expounded on the use cases of exporting a function in the latest commit.

File: README.md (hunk 12 above)
Comment (kepta) [reply README.md]: Yup, will remove that part.

File: README.md (hunk 9 above)
Comment (kepta) [reply README.md]: I have removed this line.

**Comment by kepta** [comment]
@davidtheclark updated the PR with your suggestions and also enabled returning a Promise in `underreact.config.js`.

**Pull request merged**